KAFKA-5370: Replace uses of the old consumer with the new consumer when possible

vahidhashemian opened the pull request:
Also, methods in `ClientUtils` that are called by server or tools code are introduced in `AdminUtils` with the implementation living in `AdminUtils`. All the existing callers apart from the old clients call the `AdminUtils` methods.

vahidhashemian committed:
KAFKA-5370: Replace use of old consumer with the new consumer when possible

- Uses of the old consumers in tools and tests where the new consumer would work as well (or better).

vahidhashemian commented:
@ijuma Is this close to what you had in mind by [your earlier comment](https://github.com/apache/kafka/pull/2328#issuecomment-305762105)? Thanks.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5571/
Test FAILed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5585/
Test FAILed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5577/
Test PASSed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5591/
Test FAILed (JDK 7 and Scala 2.11).

ijuma commented on core/src/test/scala/unit/kafka/admin/ListConsumerGroupTest.scala:
```diff
@@ -88,4 +88,36 @@ class ListConsumerGroupTest extends KafkaServerTestHarness {
+    val opts = new ConsumerGroupCommandOptions(Array("--bootstrap-server", brokerList))
+    val consumerGroupCommand = new KafkaConsumerGroupService(opts)
+    try {
+      assert(consumerGroupCommand.listGroups().isEmpty)
```
We should use a `JUnit` assertion (i.e. `assertTrue`).

ijuma commented on core/src/test/scala/unit/kafka/integration/TopicMetadataTest.scala:
```diff
@@ -235,58 +172,28 @@ class TopicMetadataTest extends ZooKeeperTestHarness {
-    adHocServers.last.awaitShutdown()
-
-    checkMetadata(adHocServers, numConfigs - 1)
-  }
```
Can you elaborate on why many of these tests were removed?

vahidhashemian replied:
In this PR I used an API that did not provide some of the details the deprecated API provided. In the new patch I switch to Metadata API that seem to cover the missing info. So, these tests will be back in the new patch.

ijuma commented on core/src/main/scala/kafka/admin/AdminUtils.scala:
```diff
@@ -661,9 +667,83 @@ object AdminUtils extends Logging with AdminUtilities {
+  /**
+   * Parse a list of broker urls in the form host1:port1, host2:port2, ...
+   */
+  def parseBrokerList(brokerListStr: String): Seq[BrokerEndPoint] = {
```
Maybe this should be a method in `BrokerEndPoint`.

ijuma commented on core/src/main/scala/kafka/admin/AdminUtils.scala:
```diff
@@ -661,9 +667,83 @@ object AdminUtils extends Logging with AdminUtilities {
+  def fetchTopicMetadata(topics: Set[String], brokers: Seq[BrokerEndPoint], clientId: String, timeoutMs: Int)
+    : Map[String, List[PartitionInfo]] = {
+    val props = new Properties()
+    props.put("metadata.broker.list", brokers.map(_.connectionString).mkString(","))
```
This doesn't seem to make sense since we are now using the new consumer.

vahidhashemian replied:
In the new patch this method is no longer needed.

ijuma commented on core/src/main/scala/kafka/admin/AdminUtils.scala:
```diff
@@ -661,9 +667,83 @@ object AdminUtils extends Logging with AdminUtilities {
       .map(entityPath => (entityPath, fetchEntityConfig(zkUtils, rootEntityType, entityPath))).toMap
   }
 
+  def fetchTopicMetadata(topics: Set[String], brokers: Seq[BrokerEndPoint], properties: Properties)
```
On second thought, this should probably not be here. `AdminUtils` is the place for methods that interact with `ZK`. We could add it to the `ClientUtils` in the client module maybe (that makes it clear that we are communicating with Kafka via the protocol).

vahidhashemian replied:
Do we want to deprecate `ClientUtils`? If so, should we find another host for this method? Suggestions? What do you think about `channelToAnyBroker` and `channelToOffsetManager`?

ijuma reviewed (commented):
Thanks for the PR and sorry for the delay. I left some comments. Also, it would be great if we could update `ProducerBounceTest` to use the new consumer.

ijuma commented:
One more thing: we should consider removing methods from the Scala `ClientUtils` if the new implementation is suitable across the board (like the method that parses the broker list).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6571/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6556/
Test PASSed (JDK 8 and Scala 2.12).

vahidhashemian reviewed (commented):
@ijuma Thanks for reviewing the PR. I tried to address your comments. I asked inline about where to put a couple of existing methods. Let me know if you have any suggestions. Thanks.

vahidhashemian committed:
KAFKA-5370: Replace use of old consumer with the new consumer when possible

Uses of the old consumers in tools and tests where the new consumer would work as well (or better) should be updated.

Conflicts:
	core/src/test/scala/unit/kafka/integration/TopicMetadataTest.scala

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6630/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6615/
Test PASSed (JDK 8 and Scala 2.12).

vahidhashemian commented:
@ijuma I'm making some additional adjustments. Please hold off reviewing until the next patch is submitted. Thanks.

vahidhashemian committed:
Additional cleanup

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6648/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6633/
Test PASSed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6659/
Test FAILed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6674/
Test PASSed (JDK 7 and Scala 2.11).

vahidhashemian commented:
@ijuma The PR should be in a better shape now. We may need to move the methods from `ClientUtils`, as you suggested. If so, please let me now which file(s) you think should host them.

vahidhashemian committed:
KAFKA-5370: Replace use of old consumer with the new consumer when possible

Uses of the old consumers in tools and tests where the new consumer would work as well (or better) should be updated.

Conflicts:
	core/src/test/scala/unit/kafka/integration/TopicMetadataTest.scala

Conflicts:
	core/src/test/scala/integration/kafka/api/ProducerBounceTest.scala
	core/src/test/scala/unit/kafka/integration/TopicMetadataTest.scala

ijuma commented:
Sorry, I dropped the ball on this review. I don't think this PR is relevant anymore, but I did include you as a co-author in https://github.com/apache/kafka/pull/5220 which overlaps a bit with one of the changes in this PR.
